app/code/Magento/CatalogGraphQl/Model/Resolver/Categories.php

```diff
@@ -52,68 +51,61 @@ class Categories implements ResolverInterface
     private $customAttributesFlattener;
 
     /**
-     * @var ValueFactory
-     */
-    private $valueFactory;
-
-    /**
-     * Category constructor.
      * @param CollectionFactory $collectionFactory
      * @param DataObjectProcessor $dataObjectProcessor
      * @param AttributesJoiner $attributesJoiner
      * @param CustomAttributesFlattener $customAttributesFlattener
-     * @param ValueFactory $valueFactory
      */
     public function __construct(
         CollectionFactory $collectionFactory,
         DataObjectProcessor $dataObjectProcessor,
         AttributesJoiner $attributesJoiner,
-        CustomAttributesFlattener $customAttributesFlattener,
-        ValueFactory $valueFactory
+        CustomAttributesFlattener $customAttributesFlattener
     ) {
         $this->collection = $collectionFactory->create();
         $this->dataObjectProcessor = $dataObjectProcessor;
         $this->attributesJoiner = $attributesJoiner;
         $this->customAttributesFlattener = $customAttributesFlattener;
-        $this->valueFactory = $valueFactory;
     }
 
     /**
      * {@inheritdoc}
      * @SuppressWarnings(PHPMD.UnusedFormalParameter)
      */
-    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null) : Value
+    public function resolve(Field $field, $context, ResolveInfo $info, array $value = null, array $args = null)
     {
+        if (!isset($value['model'])) {
+            throw new GraphQlInputException(__('"model" value should be specified'));
+        }
+
         /** @var \Magento\Catalog\Model\Product $product */
         $product = $value['model'];
         $categoryIds = $product->getCategoryIds();
         $this->categoryIds = array_merge($this->categoryIds, $categoryIds);
         $that = $this;
 
-        return $this->valueFactory->create(function () use ($that, $categoryIds, $info) {
-            $categories = [];
-            if (empty($that->categoryIds)) {
-                return [];
-            }
+        $categories = [];
```

Comment:
This was not supposed to be replaced because this was actual performance optimization via Deferred mechanism. 
Here `$this` serves as a registry.

Reply:
Fixed